File: src/main/java/org/lanternpowered/server/cause/entity/damage/source/AbstractEntityDamageSourceBuilder.java
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.cause.entity.damage.source;

import static com.google.common.base.Preconditions.checkNotNull;

import org.spongepowered.api.entity.Entity;
import org.spongepowered.api.event.cause.entity.damage.source.EntityDamageSource;

public abstract class AbstractEntityDamageSourceBuilder
        <T extends EntityDamageSource, B extends EntityDamageSource.EntityDamageSourceBuilder<T, B>, O extends B>
        extends AbstractDamageSourceBuilder<T, B, O> implements EntityDamageSource.EntityDamageSourceBuilder<T, B> {

    protected Entity source;

    @Override
    public O entity(Entity entity) {
        this.source = checkNotNull(entity, "Entity source cannot be null!");
        return (O) this;
    }

    @Override
    public O reset() {
        super.reset();
        this.source = null;
        return (O) this;
    }
}

File: src/main/kotlin/org/lanternpowered/server/cause/entity/damage/source/AbstractEntityDamageSourceBuilder.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.cause.entity.damage.source

import org.spongepowered.api.entity.Entity
import org.spongepowered.api.event.cause.entity.damage.source.EntityDamageSource
import org.spongepowered.api.event.cause.entity.damage.source.common.AbstractDamageSourceBuilder

@Suppress("UNCHECKED_CAST")
abstract class AbstractEntityDamageSourceBuilder<T : EntityDamageSource, B : EntityDamageSource.EntityDamageSourceBuilder<T, B>> :
        AbstractDamageSourceBuilder<T, B>(), EntityDamageSource.EntityDamageSourceBuilder<T, B> {

    internal var source: Entity? = null

    override fun entity(entity: Entity): B = apply { this.source = entity } as B

    override fun reset(): B = apply {
        super.reset()
        this.source = null
    } as B

    override fun from(value: T): B = apply {
        super.from(value)
        this.source = value.source
    } as B
}
